File: src/legimi/request.clj
(ns legimi.request
  (:require [clj-http.client :as client]
            [clojure.java.io :as io]))

(def url "https://app.legimi.pl/svc/sync/core.aspx")

(defn request
  [url body-seq]
  (client/post url
               {:body (byte-array body-seq)
                :as :byte-array}))
(defn- now
  []
  (quot (System/currentTimeMillis) 1000))

(defn- get-range
  [from len]
  [from (dec (min (+ from 81920) len))])

(defn download
  [url size filename]
  (with-open [w (io/output-stream filename)]
    (loop [[from to] (get-range 0 size)]
      (when (< from to)
        (let [book-part (client/get (format "%s&ts=%s" url (now))
                                    {:headers {"Range" (format "bytes=%d-%d" from to)}
                                     :as :byte-array})]
          (.write w (:body book-part)))
        (recur (get-range (inc to) size))))))

File: src/request.lua
local function request(url, method)
  local req = require "http.request".new_from_uri(url)
  req.setbody = function(self, body)
    self:set_body(body)
    return self
  end
  req.send = function(self)
    local _, stream = self:go()
    ---@diagnostic disable-next-line: need-check-nil
    stream.getbody = stream.get_body_as_string
    ---@diagnostic disable-next-line: need-check-nil
    stream.savetofile = stream.save_body_to_file
    return stream
  end
  if method then
    req.headers:upsert(":method", method)
  end
  return req
end

local url = "https://app.legimi.pl/svc/sync/core.aspx"

local function post(data)
  return request(url, "POST"):setbody(data):send()
end

return {
  post = post,
}
